ligo/stablecoin/actions.ligo

```diff
@@ -0,0 +1,245 @@
+  { validate_token_type (parameter.1.1.0)
+  ; const debit_params : burn_params = record
+     [ from_  = parameter.0
+     ; amount = parameter.1.1.1
+     ]
+  ; const credit_params : mint_params = record
+     [ to_    = parameter.1.0
+     ; amount = parameter.1.1.1
+     ]
+  ; const debited_storage  : storage =
+      debit_from (debit_params, accumulator)
+  ; const credited_storage : storage =
+      credit_to (credit_params, debited_storage)
+  } with credited_storage
+
+; const upd : list (operation) =
+    generic_transfer_hook (parameters, store.permissions)
+; const ups : storage =
+    List.fold (transfer_tokens, parameters, store)
+} with (upd, ups)
+
+(*
+ * Returns a list of `balance_of` calls and leaves the storage untouched
+ *)
+function balance_of
+  ( const parameter : balance_of_params
+  ; const store     : storage
+  ) : entrypoint is block
+{ function retreive_balance
+    ( const request : balance_of_request
+    ) : balance_of_response is block
+  { validate_token_type (request.token_id)
+  ; var retreived_balance : nat := 0n
+  ; case Big_map.find_opt (request.owner, store.ledger) of
+      Some (ledger_balance) -> retreived_balance := ledger_balance
+    | None -> skip
+    end
+  } with (request, retreived_balance)
+; const responses : list (balance_of_response) =
+    List.map (retreive_balance, parameter.0)
+; const transfer_operation : operation =
+    Tezos.transaction (responses, 0mutez, parameter.1)
+; const upd : list (operation) = list [transfer_operation]
+} with (upd, store)
+
+(*
+ * Retreives a list of `total_supply` that is specified for multiple
+ * token types in FA2 spec and leave the storage untouched. In reality
+ * we still restrict all of them to be of `default_token_id`.
+ *)
+function total_supply
+  ( const parameter : total_supply_params
+  ; const store     : storage
+  ) : entrypoint is block
+{ // We suppose that the token is already validated in `get_total_supply`
+  // and equal to `default_token_id` since our contract is monolitic
+  const total_supply_response_ : total_supply_response_ = record
+    [ token_id     = default_token_id
+    ; total_supply = store.total_supply
+    ]
+; const total_supply_response : total_supply_response =
+    Layout.convert_to_right_comb ((total_supply_response_ : total_supply_response_))
+  // We drop `token_id` parameter since our contract is monolitic
+  // and so we support only one `default_token_id` token type
+; function get_total_supply
+    ( const ops      : list (total_supply_response)
+    ; const token_id : token_id
+    ) : list (total_supply_response) is block
+    { validate_token_type (token_id)
+    } with total_supply_response # ops
+} with
+  ( list [Tezos.transaction
+    ( List.fold
+        ( get_total_supply
+        , parameter.0
+        , (nil : list (total_supply_response))
+        )
+    , 0mutez
+    , parameter.1
+    )]
+  , store
+  )
+
+(*
+ * Returns a list of `token_metadata` calls. This is supposed
```

Comment:
AFAIU it always returns one call. This one call can pass a list of `token_metadata` to the callback.